### PR TITLE
Working models for hubert and wav2vec2

### PR DESCRIPTION
**Pull request by florentinvonhaugwitz**
# PR Description

## Please check if the PR fulfills these requirements

- [ ] Tests for the changes have been added (for bug fixes/features) (Or not needed)
- [x] Labels, reviewers and assignees have been added to the pull request
- [x] Docs have been added / updated (for bug fixes / features) (Or not needed)

## Motivation and Context

Uploading wav2vec2 and hubert model. Only fine-tuning is still not done yet

### CODE REVIEW

File: src/stt/analyze_data.py [hunk 1]
```diff
@@ -0,0 +1,181 @@
...
+    RECORDINGS_FILE,
+    SRC_DIR,
+    STT_REPORT,
+)
+from pandas_profiling import ProfileReport
+
+sys.path.insert(0, SRC_DIR)
+from decorators import log_function_name  # noqa: E402
+
+# ! Todo
```
Comment (HannahPetry) [inline src/stt/analyze_data.py]: Brauchst du die todos noch?

File: src/stt/analyze_data.py [hunk 2]
```diff
@@ -0,0 +1,181 @@
...
+    tuple[int, float]
+        Sample rate and duration of the wav file
+    """
+    file_path = row.file_location
+    sr = librosa.get_samplerate(file_path)
+    duration = librosa.get_duration(filename=file_path)
+    return sr, np.round(duration, 3)
+
+
+@log_function_name
```
Comment (HannahPetry) [inline src/stt/analyze_data.py:49]: wofür steht immer @log_function_name über den Funktionen?

File: src/stt/evaluate_model.py [hunk 3]
```diff
@@ -0,0 +1,275 @@
...
+    batch["pred_str"] = processor.batch_decode(pred_ids)[0]
+    batch["text"] = processor.decode(
+        batch["labels"], group_tokens=False
+    )
+
+    return batch
+
+
+@log_function_name
+def show_random_elements(
```
Comment (HannahPetry) [inline src/stt/evaluate_model.py:80]: Überprüfst du mit der Funktion stichprobenartig deine results oder soll die einen anderen Nutzen haben?

File: src/stt/evaluate_model.py [hunk 4]
```diff
@@ -0,0 +1,275 @@
...
+) -> None:
+    """
+    Show random predictions and labels from a dataset
+
+    Parameters
+    ----------
+    dataset : Dataset
+        Dataset to show predictions and labels from
+    num_examples : int, optional
+        Number of predictions and labels to show, by default 10
```
Comment (HannahPetry) [inline src/stt/evaluate_model.py]: ist nicht by default 5?

File: src/stt/hubert_training.py [hunk 5]
```diff
@@ -0,0 +1,144 @@
...
+wandb.init(
+    project="speech-to-text",
+    entity="nlp_masterthesis",
+    tags=["facebook-hubert"],
+)
+
+
+@log_function_name
+def load_model(
+    processor: Wav2Vec2Processor, device: torch.device
```
Comment (HannahPetry) [inline src/stt/hubert_training.py:49]: gibt es nur den Wav2Vec2Processor für beide models?

File: src/stt/analyze_data.py (hunk 1 above)
Comment (florentinvonhaugwitz) [reply src/stt/analyze_data.py]: Guter Punkt, die sollte ich noch abhandeln bevor ich den PR merge

File: src/stt/analyze_data.py (hunk 2 above)
Comment (florentinvonhaugwitz) [reply src/stt/analyze_data.py:49]: Das erlaubt es mir anzuzeigen, welche Funktion gerade ausgeführt wird. In src/decorators.py habe ich die Funktion erstellt die das durchläuft

File: src/stt/hubert_training.py (hunk 5 above)
Comment (florentinvonhaugwitz) [reply src/stt/hubert_training.py:49]: Ja, sowohl Hubert als auch Wav2Vec2 nutzen den gleichen Processor. Deshalb muss ich die transformations auch nur einmal durchgehen

File: src/stt/evaluate_model.py (hunk 4 above)
Comment (florentinvonhaugwitz) [reply src/stt/evaluate_model.py]: Ja, gut gesehen1

File: src/stt/evaluate_model.py (hunk 3 above)
Comment (florentinvonhaugwitz) [reply src/stt/evaluate_model.py:80]: Genau, das zeigt einmal random an wie das aussehen könnte

File: src/stt/create_speech_data.py [hunk 6]
```diff
@@ -0,0 +1,184 @@
...
+    float
+        Duration of the wav file in seconds
+    """
+    return librosa.get_duration(
+        filename=os.path.join(RAW_RECORDINGS_DIR, file_path)
+    )
+
+
+@log_function_name
+def copy_files(files_list: list[str], folder: str) -> None:
```
Comment (Tara-Sophia) [inline src/stt/create_speech_data.py:52]: why do you need to copy files

File: src/stt/create_speech_data.py [hunk 7]
```diff
@@ -0,0 +1,184 @@
...
+    Returns
+    -------
+    pd.DataFrame
+        Cleaned dataframe
+    """
+
+    df["path"] = df["file_name"].apply(
+        lambda x: os.path.join(RAW_DATA_DIR, folder, x)
+    )
+    df["audio"] = df["path"]
```
Comment (Tara-Sophia) [inline src/stt/create_speech_data.py:97]: wieso ist df audio = df path

Comment (Tara-Sophia) [inline src/stt/create_speech_data.py:97]: wieso ist df audio = df path ?

Comment (Tara-Sophia) [inline src/stt/create_speech_data.py:97]: wieso ist df audio = df path

Comment (Tara-Sophia) [inline src/stt/create_speech_data.py:97]: wieso ist df audio = df path ?

File: src/stt/create_speech_data.py [hunk 8]
```diff
@@ -0,0 +1,184 @@
...
+    folder : str
+        Name of the folder to copy the files to
+
+    Returns
+    -------
+    pd.DataFrame
+        Cleaned dataframe
+    """
+
+    df["path"] = df["file_name"].apply(
```
Comment (Tara-Sophia) [inline src/stt/create_speech_data.py:94]: wieso ist df audio = df path ?

File: src/stt/evaluate_model.py [hunk 9]
```diff
@@ -0,0 +1,275 @@
...
+    Returns
+    -------
+    Example
+        Batch with results
+    """
+    with torch.no_grad():
+        input_values = torch.tensor(
+            batch["input_values"], device=device
+        ).unsqueeze(0)
+        logits = model(input_values).logits
```
Comment (Tara-Sophia) [inline src/stt/evaluate_model.py:68]: ist dein model jetzt ein model oder eine union von beiden?

File: src/stt/predict.py [hunk 10]
```diff
@@ -0,0 +1,113 @@
...
+    Transcribe audio from a microphone
+
+    Parameters
+    ----------
+    pipe : Pipeline
+        Speech recognition pipeline
+    rec : sr.Recognizer
+        Speech recognition recognizer
+    """
+    with sr.Microphone(sample_rate=16000) as source:
```
Comment (Tara-Sophia) [inline src/stt/predict.py:42]: was ist Mircrophone eine function?

File: src/stt/streamlit_stt.py [hunk 11]
```diff
@@ -0,0 +1,132 @@
...
+    tuple(Wav2Vec2ForCTC, Wav2Vec2Processor)
+        Trained wav2vec2 model and processor
+    """
+    model = Wav2Vec2ForCTC.from_pretrained(WAV2VEC2_MODEL_DIR)
+    processor = Wav2Vec2Processor.from_pretrained(VOCAB_DIR)
+    model.to(device)
+    return model, processor
+
+
+# FACEBOOK HUBERT
```
Comment (Tara-Sophia) [inline src/stt/streamlit_stt.py:68]: raus oder noch to do?

File: src/stt/streamlit_stt.py [hunk 12]
```diff
@@ -0,0 +1,132 @@
...
+#     tuple(HubertForCTC, Wav2Vec2Processor)
+#         Trained hubert model and processor
+#     """
+#     model = HubertForCTC.from_pretrained(HUBERT_MODEL_DIR)
+#     processor = Wav2Vec2Processor.from_pretrained(VOCAB_DIR)
+#     model.to(device)
+#     return model, processor
+
+
+st.sidebar.title("Choose the model")
```
Comment (Tara-Sophia) [inline src/stt/streamlit_stt.py:92]: müssen wir uns hier vielleicht auch einheitlich überlegen, ob wir dem nutzer nur das beste model als default geben wollen und er sich die anderen noch aussuchen kann oder nicht

File: src/stt/transform_speech_data.py [hunk 13]
```diff
@@ -0,0 +1,370 @@
...
+            Dictionary containing all characters from the batch
+        """
+        all_text = " ".join(batch["sentence"])
+        vocab = list(set(all_text))
+        return {"vocab": [vocab], "all_text": [all_text]}
+
+    vocab_train = train_ds.map(
+        extract_all_chars,
+        batched=True,
+        batch_size=-1,
```
Comment (Tara-Sophia) [inline src/stt/transform_speech_data.py:139]: batched true to size 1 - macht das sinn? oder kann man dann direkt batch false sagen

File: src/stt/streamlit_stt.py (hunk 12 above)
Comment (florentinvonhaugwitz) [reply src/stt/streamlit_stt.py:92]: Ja, auf jeden Fall. Sollte in der finalen Version nicht mehr vorhanden sein

File: src/stt/streamlit_stt.py (hunk 11 above)
Comment (florentinvonhaugwitz) [reply src/stt/streamlit_stt.py:68]: Fürs erste bleibt das drin. Das nehme ich dann aber bei Gelegenheit wieder raus

File: src/stt/predict.py (hunk 10 above)
Comment (florentinvonhaugwitz) [reply src/stt/predict.py:42]: Ne das ist eine class

File: src/stt/evaluate_model.py (hunk 9 above)
Comment (florentinvonhaugwitz) [reply src/stt/evaluate_model.py:68]: Ist jeweils nur ein Model. Union bedeuter, das entweder das eine oder das andere Model als input gegeben werden kann

File: src/stt/create_speech_data.py (hunk 8 above)
Comment (florentinvonhaugwitz) [reply src/stt/create_speech_data.py:94]: Das liegt daran, wie man das leider für ein Audio dataset machen muss. Audio ist immer der path zur Audio datei

File: src/stt/create_speech_data.py (hunk 6 above)
Comment (florentinvonhaugwitz) [reply src/stt/create_speech_data.py:52]: Dann habe ich sie immer an beiden Orten für den Fall das ich etwas an der Funktion verändern will

File: src/stt/create_speech_data.py (hunk 7 above)
Comment (florentinvonhaugwitz) [reply src/stt/create_speech_data.py:97]: Das liegt daran wir eine Audio dataset aufgebaut sein muss. Das muss quasi der genaue Pfad zur Audio datei sein

File: src/stt/transform_speech_data.py (hunk 13 above)
Comment (florentinvonhaugwitz) [reply src/stt/transform_speech_data.py:139]: Das gibt quasi die bestmögliche batch size an

**Review by HannahPetry — APPROVED** [review]
